<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Barman, makes you your G&amp;T (plots) â€¢ barman</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Barman, makes you your G&amp;T (plots)">
<meta property="og:description" content="Convenient functions to aid in pre-processing and plotting multiomic data produced with the G&amp;T-seq method.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">barman</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/seanlaidlaw/barman/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="barman" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#barman" class="anchor"></a>Barman</h1></div>
<p><img src="https://github.com/seanlaidlaw/barman/blob/master/img/barman_logo.png?raw=true"></p>
<p>Barman is an R library for easy data preprocessing and visualization of Genome and Transcriptome single-cell data.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Barman is still under active development, as such installation is done via the actively maintained <a href="https://github.com/seanlaidlaw/barman">git repository</a></p>
<pre><code><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">devtools::install_github("seanlaidlaw/barman")</a></code></pre>
</div>
<div id="standard-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#standard-workflow" class="anchor"></a>Standard workflow</h2>
<p>Starting from a gene x cell counts matrix barman functions can be used in the following order to achieve a standard workflow:</p>
<pre><code>                +------------------------------+
                |       RNA Counts Matrix      |
                +--------------+---------------+
                               |
                               |
                               |                           +-----+
                               v                                 |
                +--------------+---------------+                 |
                | filter_and_normalise_scrna() |                 |
                +--------------+---------------+                 +
                               |                              counts
                               v                           normalisation
                +--------------+---------------+                 +
                |      logR_by_ref_group()     |                 |
                +---------+---------+----------+                 |
                          |         |                            |
                          |         |                      +-----+
                          |   OR    |
                          |         |                      +-----+
                          v         v                            |
+-------------------------+-+    +--+------------------------+   |
|   expression_boxplots()   |    |get_expression_by_segment()|   |
+---------------------------+    +-------------+-------------+   |
                                               |                 |
                                               |           visualisation
                                               v                 |
                                 +-------------+-------------+   |
                                 |       G_T_chr_plot()      |   |
                                 +---------------------------+   |
                                                                 |
                                                           +-----+</code></pre>
</div>
<div id="quick-start" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h2>
<p>Barman is designed for multiomic data, but provides functions for preprocessing DNA and RNA independently.</p>
<p>Preprocessing the RNA requires a counts matrix, which can be generated from single-cell RNA bams using a tool such as <a href="dx.doi.org/10.1093/bioinformatics/btt656">FeatureCounts</a>.</p>
<pre><code># Load the counts
counts_matrix = read.table("my_featurecounts_matrix.tsv", sep = "\t", header = TRUE)</code></pre>
<p>Filtering can be done automatically using PCA outlier detection, and the resulting filtered matrix is noramlized for gene length using FPKM.</p>
<pre><code># Filtering and Normalizing
counts_matrix = filter_and_normalise_scrna(counts_matrix) # automatic filtering</code></pre>
<p>Filtering can also be manually specified using the <code>manual_filter</code> option and providing a list of 2 length vectors, specifying the lower and upper limits for <code>total_counts</code>, <code>total_features_by_counts</code>, %MT genes, and %ERCC respectively.</p>
<pre><code># Filtering and Normalizing
fpkm_matrix = filter_and_normalise_scrna(counts_matrix, manual_filter = list(c(100000,100000000),
c(1000, 5000), 20, 50)) # filters out cells according to manual filters</code></pre>
<p>We can see the difference in the QC by running the <code>qc_plots</code> function</p>
<pre><code><a href="reference/qc_plots.html">qc_plots(fpkm_matrix)</a></code></pre>
<p>We can also normalize based on a given reference group by using the <code>logR_by_ref_group</code> function to calculate the log fold change between a group of specified reference cells and the rest:</p>
<pre><code>reference_group_cells = rownames(fpkm_matrix)[fpkm_matrix$cell_type = "control",]
logFC_matrix = logR_by_ref_group(fpkm_matrix, reference_group_cells)</code></pre>
<p>We can plot expression per chromosome of different cells by using the <code>expression_boxplots</code> function which accepts two lists of cell ids, one as an experimental group, and one as a control. From this it calcualtes the average expression per gene, and groupes by chromosome before plotting.</p>
<pre><code>expression_boxplots(experimental_group = list("Cell_A01", "Cell_A02", "Cell_B03"), control_group =
list("Cell_D03", "Cell_D04", "Cell_D05", fpkm_matrix)</code></pre>
<p>If scCNV has been run on the DNA bams then we can use the produced segmentation files, and a counts matrix to produce segmentation files for the RNA expression data, thus allowing us to compare segment to segment between DNA and RNA.</p>
<pre><code>scCNV_segment_for_A01 = read.table("SEGMENT_scCNV_segment_for_A01.copynumber.refLocus.txt"
A01_expression_segments = get_expression_by_segment("Cell_A01", fpkm_matrix, scCNV_segment_for_A01)</code></pre>
<p>This can also be done in bulk, without having to read in the scCNV produced segment files by using the <code>bulk_get_expression_by_segment</code> function. this function uses all available cores except one to process the files in parallel.</p>
<pre><code><a href="reference/bulk_get_expression_by_segment.html">bulk_get_expression_by_segment(fpkm_matrix, "./Path_to_scCNV_outputs/", "./RNA_seg_output_folder")</a></code></pre>
<p>Finally, we can see the Genome and Transcriptome segment plot, showing the DNA to RNA comparison by running the <code>G_T_chr_plot</code> function.</p>
<pre><code>G_T_chr_plot(cnv_data = scCNV_segment_for_A01, exp_data = A01_expression_segments, title = "Cell A01 Genome and Transcriptome")</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/seanlaidlaw/barman/">https://â€‹github.com/â€‹seanlaidlaw/â€‹barman/â€‹</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPLv3</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://seanlaidlaw.com">Sean Laidlaw</a> <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://seanlaidlaw.com">Sean Laidlaw</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
