#' @title barman
#'
#' @description Generates Genome CNV and Transcriptome Expression plots by chromsome from previously generated segmentation files.
#'
#' @param cnv_data a segmentation file produced by scCNV. Can also use an ASCAT.summary.csv if converted with the ascat2sccnvseg() function.
#' @param exp_data a segmentation file generated by calling the get_expression_by_segment() function on a featurecounts matrix and on a DNA segmentation file
#' @param title Plot title
#' @param save boolean to specify if the plot should be generated or saved as PNG. If set to TRUE, the plot will be saved as the specified title.
#' @param highlight_gene name of a gene to show in red on G_T_chr_plot, allows visualisation of copy number for a specific gene
#'
#' @return a ggplot2 object or nothing if save==TRUE
#'
get_copynumber_for_gene = function(dna_segments, chr, start, end, gene_name) {

	if (!missing(gene_name)) {
	mart72.hs <- biomaRt::useMart("ENSEMBL_MART_ENSEMBL", "hsapiens_gene_ensembl", host = "grch37.ensembl.org")

		if (grepl("ENSG[0-9]*", gene_name)) {
			new_gene_ids = biomaRt::getBM(
				attributes=c("ensembl_gene_id","chromosome_name","start_position","end_position"),
				filters="ensembl_gene_id",
				values=gene_name,
				mart=mart72.hs)
		} else {
			new_gene_ids = biomaRt::getBM(
				attributes=c("hgnc_symbol","chromosome_name","start_position","end_position"),
				filters="hgnc_symbol",
				values=gene_name,
				mart=mart72.hs)

		}
		if (nrow(new_gene_ids) < 1) {
			print(paste0("Error: No matches found on biomaRt for gene_name:", gene_name))
		}

		chr = new_gene_ids$chromosome_name
		start = new_gene_ids$start_position
		end = new_gene_ids$end_position
	} else {
		# if no gene+name given make sure that chr start and end are given
		if (missing(chr) | missing(start) | missing(end)) {
			print("Error: Please either provide gene_name arg or chr,start,end arguments")
			return(NA)
		}
	}


	# look for segments that entirely contain gene
	n_entire_containing_segments = dna_segments[(dna_segments$Chr == as.character(chr) & dna_segments$Pos <= as.integer(start) & dna_segments$end >=as.integer(end)),]

	if (nrow(n_entire_containing_segments) > 0) {
		segStart = n_entire_containing_segments$Pos
		segEnd = n_entire_containing_segments$end
		segraw_1 = n_entire_containing_segments$rawCN
		segraw_2 = n_entire_containing_segments$rawCN
		segsegCN_1 = n_entire_containing_segments$segmentedCN
		segsegCN_2 = n_entire_containing_segments$segmentedCN
	} else {
		# if gene is spread over multiple segments look for the closest near start
		n_start_containing_segments = dna_segments[(dna_segments$Chr == as.character(chr)),]
		n_start_containing_segments = n_start_containing_segments[order(n_start_containing_segments$Pos),]
		n_start_containing_segments = tail(n_start_containing_segments[(n_start_containing_segments$Pos <= as.integer(start)),], 1)
		segStart = n_start_containing_segments$Pos
		segraw_1 = n_start_containing_segments$rawCN
		segsegCN_1 = n_start_containing_segments$segmentedCN

		n_end_containing_segments = dna_segments[(dna_segments$Chr == as.character(chr)),]
		n_end_containing_segments = n_end_containing_segments[order(n_end_containing_segments$end),]
		n_end_containing_segments = head(n_end_containing_segments[(n_end_containing_segments$end >= as.integer(end)),], 1)
		segEnd = n_end_containing_segments$end
		segraw_2 = n_end_containing_segments$rawCN
		segsegCN_2 = n_end_containing_segments$segmentedCN
	}

	return_values = list(chr, segStart, segEnd, segraw_1, segraw_2, segsegCN_1, segsegCN_2)
	return(return_values)
}


#' @export
G_T_chr_plot <- function(cnv_data, exp_data, highlight_gene, title, save) {
	if (missing(save)) {
		save = FALSE
	}

	if (missing(title)) {
		title = ""
	}

	if (!missing(highlight_gene)) {
		highlight_genes_info = get_copynumber_for_gene(cnv_data, gene_name = highlight_gene)

	}


	cnv_data$Chr <- paste0("chr", cnv_data$Chr)
	cnv_data_upper_cutoff <- round(mean(cnv_data$segmentedCN) + sd(cnv_data$segmentedCN) * 3)
	cnv_data_upper_cutoff = 4
	cnv_data <- cnv_data[cnv_data$rawCN <= cnv_data_upper_cutoff, ]

	exp_data$Chr <- paste0("chr", exp_data$Chr)
	exp_data_upper_cutoff <- mean(exp_data$PCFexp) + signif(sd(exp_data$SegmentExp) * 3, digits = 2)
	exp_data <- exp_data[exp_data$SegmentExp <= exp_data_upper_cutoff, ]


	dna_r0 <- 0
	dna_r1 <- 0.9
	rna_r0 <- 0
	rna_r1 <- 0.9

	if (save) {
		png(paste0(title, ".png"), width = 5120, height = 1600, units = "px", res = 300)
	}

	# TODO: add kpAddLabels() for xlab and ylabs and add kpAddMainTitle() for main title
	kp <- karyoploteR::plotKaryotype(genome = "hg19", plot.type = 3, cex = 0.85, main = gsub(".*__", "", title))
	karyoploteR::kpDataBackground(kp, data.panel = 2, r0 = dna_r1, r1 = dna_r0)
	karyoploteR::kpAxis(karyoplot = kp, col = "gray50", data.panel = 2, ymin = 0, ymax = cnv_data_upper_cutoff, r1 = dna_r0, r0 = dna_r1, cex = 0.5, numticks = 10, side = 2)
	karyoploteR::kpPoints(kp, chr = cnv_data$Chr, x = cnv_data$Pos, y = cnv_data$rawCN, col = "#00A6EDAA", data.panel = 2, ymin = 0, ymax = cnv_data_upper_cutoff, r0 = dna_r1, r1 = dna_r0, clipping = T)
	karyoploteR::kpSegments(karyoplot = kp, data.panel = 2, chr = cnv_data$Chr, x0 = cnv_data$Pos, x1 = cnv_data$end, y0 = cnv_data$segmentedCN, y1 = cnv_data$segmentedCN, ymin = 0, ymax = cnv_data_upper_cutoff, r0 = dna_r1, r1 = dna_r0, clipping = T, col = "black", lwd = 3)
	if (!missing(highlight_gene)) {
		if (length(highlight_genes_info) == 7) {
			# if all the columns we expect are there
				mychr = paste0("chr",highlight_genes_info[[1]])
				myx0 = highlight_genes_info[[2]]
				myx1 = highlight_genes_info[[3]]

				myy0 = highlight_genes_info[[4]]
				myy1 = highlight_genes_info[[5]]
				myY = mean(myy0, myy1)

			karyoploteR::kpSegments(karyoplot = kp, data.panel = 2, chr = mychr, x0 = myx0, x1 = myx1, y0 = myy0, y1 = myy1, ymin = 0, ymax = cnv_data_upper_cutoff, r0 = dna_r1, r1 = dna_r0, clipping = T, col = "#EB7A86", lwd = 5)
		}
	}


	karyoploteR::kpText(karyoplot = kp, data.panel = 2, chr = "chr1", y = 0.283, x = 0, labels = "Copy Number\n\n", srt = 90, cex = 0.8)

	karyoploteR::kpDataBackground(kp, data.panel = 1, r0 = rna_r0, r1 = rna_r1)
	karyoploteR::kpAxis(karyoplot = kp, col = "gray50", data.panel = 1, ymin = round(min(exp_data$SegmentExp)), ymax = exp_data_upper_cutoff, r1 = rna_r1, r0 = rna_r0, cex = 0.5, numticks = 10, side = 2)
	karyoploteR::kpPoints(kp, chr = exp_data$Chr, x = exp_data$Start, y = exp_data$SegmentExp, col = "#FFBD07AA", data.panel = 1, ymin = round(min(exp_data$SegmentExp)), ymax = exp_data_upper_cutoff, r0 = rna_r0, r1 = rna_r1, clipping = T)
	karyoploteR::kpSegments(karyoplot = kp, data.panel = 1, chr = exp_data$Chr, x0 = exp_data$Start, x1 = exp_data$End, y0 = exp_data$PCFexp, y1 = exp_data$PCFexp, ymin = round(min(exp_data$SegmentExp)), ymax = exp_data_upper_cutoff, r0 = rna_r0, r1 = rna_r1, clipping = T, col = "black", lwd = 3)
	karyoploteR::kpText(karyoplot = kp, data.panel = 1, chr = "chr1", y = 0.3, x = 0, labels = "Segmented Expression\n\n", srt = 90, cex = 0.8)

	if (save) {
		dev.off()
	}
}

